### PR TITLE
Use tomlkit for version bumps

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- preserve Cargo.toml formatting and comments during version bumps
- update ortho_config crate's ortho_config_macros dependency when bumping versions

## Testing
- `make fmt`
- `make lint`
- `make test`

------
https://chatgpt.com/codex/tasks/task_e_68ad64a5f790832283d93418713d8f29

## Summary by Sourcery

Use tomlkit in the version bump script to maintain Cargo.toml formatting and comments and add support for synchronising internal dependency versions when bumping crate versions

Enhancements:
- Switch bump_version.py to tomlkit for parsing and writing Cargo.toml to preserve formatting and comments
- Add optional dependency parameter to synchronise internal crate dependency versions (used for ortho_config_macros)

### CODE REVIEW

File: scripts/bump_version.py [hunk 1]
```diff
@@ -20,19 +20,48 @@
...
-    with toml_path.open("rb") as fh:
-        data = tomllib.load(fh)
-    if "workspace" in data and "package" in data["workspace"]:
-        data["workspace"]["package"]["version"] = version
-    elif "package" in data:
-        data["package"]["version"] = version
+def _set_version(
+    toml_path: Path, version: str, dependency: str | None = None
+) -> None:
+    """Set version fields in a ``Cargo.toml`` file.
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:29]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: scripts/bump_version.py [hunk 2]
```diff
@@ -230,9 +262,9 @@ def main(argv: list[str]) -> int:
...
     workspace = root / "Cargo.toml"
     try:
-        with workspace.open("rb") as fh:
-            data = tomllib.load(fh)
-    except tomllib.TOMLDecodeError as exc:  # pragma: no cover - defensive
+        with workspace.open("r", encoding="utf-8") as fh:
+            data = tomlkit.parse(fh.read())
+    except TOMLKitError as exc:  # pragma: no cover - defensive
         print(f"Error: Failed to parse {workspace}: {exc}", file=sys.stderr)
         return 1
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:445]: Has this now been resolved?

File: scripts/bump_version.py [hunk 3]
```diff
@@ -20,19 +20,48 @@
...
 
-def _set_version(toml_path: Path, version: str) -> None:
-    """Set the package or workspace version in a Cargo.toml file."""
-    with toml_path.open("rb") as fh:
-        data = tomllib.load(fh)
-    if "workspace" in data and "package" in data["workspace"]:
-        data["workspace"]["package"]["version"] = version
-    elif "package" in data:
-        data["package"]["version"] = version
+def _set_version(
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:138]: Has this now been resolved?

File: scripts/bump_version.py [hunk 4]
```diff
@@ -19,20 +20,171 @@
...
+            prefix = (
+                entry.value[0]
+                if entry.value and entry.value[0] in "^~"
+                else ""
+            )
+            entry.value = prefix + version
+        else:
+            text = str(entry)
+            prefix = text[0] if text and text[0] in "^~" else ""
+            deps[dependency] = prefix + version
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:135]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:135]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: scripts/bump_version.py [hunk 5]
```diff
@@ -20,19 +20,48 @@
...
+
+    if dependency:
+        deps = doc.get("dependencies")
+        if deps and dependency in deps:
+            entry = deps[dependency]
+            if isinstance(entry, dict):
+                entry["version"] = version
+            else:
+                deps[dependency] = version
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py]: Has this now been resolved?

File: scripts/bump_version.py [hunk 6]
```diff
@@ -1,12 +1,13 @@
 #!/usr/bin/env -S uv run
 # /// script
-# dependencies = ["tomli-w"]
+# dependencies = ["tomlkit", "markdown-it-py"]
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Has the following now been resolved?

❌ NOT RESOLVED: Tests for dependency formats

No dedicated test files exist for the version bumping functionality. While the functions contain doctest examples, there are no comprehensive tests covering:

    Dependencies expressed as simple strings vs tables
    Preservation of formatting and version prefixes (caret/tildes)
    Edge cases and error conditions

**Comment by leynos** [comment]
@coderabbitai pause

File: scripts/tests/test_bump_version.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+```
+post
+"""
+    for rel in ("README.md", "docs/users-guide.md"):
+        md_path = tmp_path / rel
+        md_path.parent.mkdir(parents=True, exist_ok=True)
+        md_path.write_text(md_text)
+        _update_markdown_versions(md_path, "1")
+        updated = md_path.read_text()
+        assert 'ortho_config = "1"' in updated, "must update TOML fences"
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/tests/test_bump_version.py:76]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai pause

File: scripts/tests/test_bump_version.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    md_text = (
+        "pre\n" "```toml\n" "[dependencies]\n" 'ortho_config = "0"\n' "```\n" "post\n"
+    )
+    for rel in ("README.md", "docs/users-guide.md"):
+        md_path = tmp_path / rel
+        md_path.parent.mkdir(parents=True, exist_ok=True)
+        md_path.write_text(md_text)
+        _update_markdown_versions(md_path, "1")
+        updated = md_path.read_text()
+        assert 'ortho_config = "1"' in updated, "must update TOML fences"
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/tests/test_bump_version.py]: Has this now been resolved?

File: scripts/tests/test_bump_version.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    dumped = tomlkit.dumps(doc)
+    assert 'version = "~1.2.3"' in dumped, "should update dict dependency version"
+    assert 'features = ["a"]' in dumped, "should preserve other fields"
+
+
+def test_updates_dev_dependency_table() -> None:
+    doc = tomlkit.parse('[dev-dependencies]\nfoo = "0.1"')
+    _update_dependency_version(doc, 'foo', '1.2.3')
+    assert 'foo = "1.2.3"' in tomlkit.dumps(doc), "should update dev-dependencies table"
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/tests/test_bump_version.py]: Has this now been resolved?

File: scripts/bump_version.py [hunk 10]
```diff
@@ -18,21 +19,236 @@
...
+        entry["version"] = prefix + version
+
+
+def _update_string_dependency(
+    deps: MutableMapping[str, object],
+    dependency: str,
+    entry: Any,
+    version: str,
+) -> None:
+    """Update string-style dependency ``dependency`` in ``deps``.
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py]: Has this now been resolved?

File: scripts/bump_version.py [hunk 11]
```diff
@@ -18,21 +19,236 @@
...
+            continue
+        start, end = tok.map
+        out.append("".join(lines[last:start]))
+        fence_marker = tok.markup or "```"
+        info = tok.info or lang
+        new_body = replace_fn(tok.content)
+        out.append(f"{fence_marker}{info}\n{new_body}\n{fence_marker}\n")
+        last = end
+    out.append("".join(lines[last:]))
+    return "".join(out)
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py]: Has this now been resolved?

File: scripts/bump_version.py [hunk 12]
```diff
@@ -18,21 +19,236 @@
...
+    if isinstance(existing, tomlkit.items.String):
+        new = tomlkit.string(prefix + version)
+        new.trivia.indent = existing.trivia.indent
+        new.trivia.comment_ws = existing.trivia.comment_ws
+        new.trivia.comment = existing.trivia.comment
+        new.trivia.trail = existing.trivia.trail
+        entry["version"] = new
+    else:
+        entry["version"] = prefix + version
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py]: @codex has this issue been resolved?

File: scripts/tests/test_bump_version.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/tests/test_bump_version.py]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: scripts/tests/test_bump_version.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    _update_dependency_version(doc, 'foo', '1.2.3')
+    assert 'foo = "^1.2.3"' in tomlkit.dumps(doc), "should update build-dependencies table"
+
+
+def test_preserves_trailing_comment_on_string_dependency() -> None:
+    doc = tomlkit.parse('[dependencies]\nfoo = "^0.1"  # pinned for CI\n')
+    _update_dependency_version(doc, 'foo', '1.2.3')
+    dumped = tomlkit.dumps(doc)
+    assert '# pinned for CI' in dumped, "must preserve trailing comment on value node"
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/tests/test_bump_version.py:44]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: scripts/tests/test_bump_version.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
+import tomlkit
+from scripts.bump_version import (
+    _update_dependency_version,
+    _update_markdown_versions,
+)
+
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/tests/test_bump_version.py]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: scripts/bump_version.py [hunk 15]
```diff
@@ -203,6 +422,41 @@
...
+    ...     fh.seek(0)
+    ...     'ortho_config = "1"' in fh.read()
+    True
+    """
+    if not md_path.exists():
+        return
+    text = md_path.read_text(encoding="utf-8")
+    updated = replace_fences(text, "toml", lambda body: _replace_version_in_toml(body, version))
+    md_path.write_text(updated, encoding="utf-8")
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:504]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: scripts/bump_version.py [hunk 16]
```diff
@@ -203,6 +422,41 @@
...
+    try:
+        doc = tomlkit.parse(snippet)
+    except TOMLKitError:
+        return snippet
+    for table in ("dependencies", "dev-dependencies", "build-dependencies"):
+        deps = doc.get(table)
+        if deps and "ortho_config" in deps:
+            _update_dependency_in_table(deps, "ortho_config", version)
+    return tomlkit.dumps(doc).rstrip()
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:482]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: scripts/bump_version.py [hunk 17]
```diff
@@ -18,21 +19,236 @@
...
+    >>> _update_dependency_version(doc, 'foo', '1.2.3')
+    >>> 'foo = "^1.2.3"' in tomlkit.dumps(doc)
+    True
+    """
+    for table in ("dependencies", "dev-dependencies", "build-dependencies"):
+        deps = doc.get(table)
+        if not deps or dependency not in deps:
+            continue
+        _update_dependency_in_table(deps, dependency, version)
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply scripts/bump_version.py:267]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.
